File: plugins/capacitor-kakao-login/android/src/main/java/hanhokim/capacitor/kakao/login/CapacitorKakaoLoginPlugin.java
package hanhokim.capacitor.kakao.login;

import android.app.Activity;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.util.Log;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import com.kakao.sdk.auth.model.OAuthToken;
import com.kakao.sdk.common.KakaoSdk;
import com.kakao.sdk.user.UserApiClient;

import org.json.JSONException;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;


@CapacitorPlugin(name = "CapacitorKakaoLogin")
public class CapacitorKakaoLoginPlugin extends Plugin {

    private CapacitorKakaoLogin implementation = new CapacitorKakaoLogin();

    @Override
    public void load() {
        super.load();

        try {
            // capacitor.config.ts에서 kakaoAppKey 가져오기
            String kakaoAppKey = getConfig().getString("app_key", null);

            // 플러그인 로드 시 Kakao SDK 자동 초기화
            if (kakaoAppKey != null && !kakaoAppKey.isEmpty()) {
                KakaoSdk.init(getContext(), kakaoAppKey);
                Log.d("Capacitor", "✅ Kakao SDK initialized with App Key: " + kakaoAppKey);
            } else {
                Log.e("Capacitor", "❌ Kakao SDK initialization failed: App Key is missing.");
            }
        } catch (Exception e) {
            Log.e("Capacitor", "❌ Kakao SDK initialization error: " + e.getMessage(), e);
        }
    }

    // JS 에서 호출할 수 있는 빈 initSDK() 함수. 이게 실행되면 load() 가 실행되면서 initSDK 가 실행된다.
    @PluginMethod
    public void initSDK(PluginCall call) {
        call.resolve();
    }

    @PluginMethod
    public void echo(PluginCall call) {
        String value = call.getString("value");

        JSObject ret = new JSObject();
        ret.put("value", implementation.echo(value));
        call.resolve(ret);
    }

    @PluginMethod
    public void prompt(PluginCall call) {
        Activity activity = getActivity();

        if (UserApiClient.getInstance() == null) {
            call.reject("Kakao SDK가 초기화되지 않았습니다.");
            return;
        }

        // JS에서 넘어온 scopes 배열 읽기
        JSArray jsScopes = call.getArray("scopes");
        List<String> scopes = new ArrayList<>();
        if (jsScopes != null) {
            for (int i = 0; i < jsScopes.length(); i++) {
                try {
                    scopes.add(jsScopes.getString(i));
                } catch (JSONException e) {
                    call.reject("Invalid scope value at index " + i);
                    return;
                }
            }
        }

        // KakaoTalk 앱이 설치되어 있는지 확인
        if (UserApiClient.getInstance().isKakaoTalkLoginAvailable(activity)) {
            // KakaoTalk 앱으로 로그인
            UserApiClient.getInstance().loginWithKakaoTalk(activity, (oauthToken, error) -> {
                handleLoginResult(oauthToken, error);
                return null;
            });
        } else {
            // Kakao 계정 웹 로그인
            UserApiClient.getInstance().loginWithKakaoAccount(activity, (oauthToken, error) -> {
                handleLoginResult(oauthToken, error);
                return null;
            });
        }

        call.resolve();
    }

    private void handleLoginResult(OAuthToken oauthToken, Throwable error) {
        if (error != null) {
            Log.e("Capacitor", "❌ Kakao login failed", error);
            JSObject errorData = new JSObject();
            errorData.put("success", false);
            errorData.put("error", error.getLocalizedMessage());
            notifyListeners("callback", errorData);
        } else if (oauthToken != null) {
            Log.d("Capacitor", "✅ Kakao login successful: " + oauthToken.getAccessToken());

            // iOS와 동일한 데이터 구조 반환 (callback 이벤트)
            JSObject response = new JSObject();
            response.put("success", true);
            response.put("access_token", oauthToken.getAccessToken());
            response.put("expires_in", ""); // Not valid in Android
            response.put("refresh_token", oauthToken.getRefreshToken());
            response.put("refresh_token_expires_in", ""); // Not valid in Android

            // `refreshTokenExpiredAt`는 `Date` 타입이므로 변환 필요
            Date refreshTokenExpiredAt = oauthToken.getRefreshTokenExpiresAt();
            response.put("refresh_token_expired_at", refreshTokenExpiredAt != null ? refreshTokenExpiredAt.getTime() : null);

            response.put("id_token", oauthToken.getIdToken() != null ? oauthToken.getIdToken() : ""); // Can be null
            response.put("token_type", "Bearer");

            // callback 이벤트로 전송
            notifyListeners("callback", response);
        }
    }

}

File: plugins/capacitor-kakao-login/ios/Sources/CapacitorKakaoLoginPlugin/CapacitorKakaoLoginPlugin.swift
import Foundation
import KakaoSDKUser
import KakaoSDKCommon
import KakaoSDKAuth
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CapacitorKakaoLoginPlugin)
public class CapacitorKakaoLoginPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapacitorKakaoLoginPlugin"
    public let jsName = "CapacitorKakaoLogin"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "echo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "initSDK", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "prompt", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = CapacitorKakaoLogin()

    override public func load() {
        super.load()

        let pluginConfig = bridge?.config.getPluginConfig("CapacitorKakaoLogin")

        if let kakaoAppKey = pluginConfig?.getString("app_key") {
            KakaoSDK.initSDK(appKey: kakaoAppKey)
            print("✅ Kakao SDK initialized with App Key: \(kakaoAppKey)")
        } else {
            print("❌ Kakao SDK initialization failed: App Key is missing in capacitor.config.ts")
        }
    }

    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve([
            "value": implementation.echo(value)
        ])
    }

    @objc func initSDK(_ call: CAPPluginCall) {
        // 아무것도 하지 않는다. 다만 최초 init()이 실행되면 load()가 실행될 것이기 때문임
        call.resolve()
    }

    @objc func prompt(_ call: CAPPluginCall) {
        let scopes = call.getArray("scopes", String.self) ?? []

        if UserApi.isKakaoTalkLoginAvailable() {
            // 카카오톡 앱 로그인
            // handleLoginResult이 실행되려면 AppDelegation.swift에 코드가 추가되어야함 (README 참고)
          UserApi.shared.loginWithKakaoTalk() { (oauthToken, error) in
                self.handleLoginResult(oauthToken, error, call)
            }
        } else {
            // 카카오 계정 웹 로그인
            // handleLoginResult이 실행되려면 AppDelegation.swift에 코드가 추가되어야함 (README 참고)
            UserApi.shared.loginWithKakaoAccount(scopes: scopes) { (oauthToken, error) in
                self.handleLoginResult(oauthToken, error, call)
            }
        }
        call.resolve()
    }

    private func handleLoginResult(_ oauthToken: OAuthToken?, _ error: Error?, _ call: CAPPluginCall) {
        if let error = error {
            print("❌ Kakao login failed: \(error.localizedDescription)")
            self.notifyListeners("callback", data: ["success": false, "error": error.localizedDescription])
        } else if let oauthToken = oauthToken {
            print("✅ Kakao login successful. Token saved.")

            self.notifyListeners("callback", data: [
                "success": true,
                "access_token": oauthToken.accessToken,
                "expires_in": oauthToken.expiresIn,
                "refresh_token": oauthToken.refreshToken,
                "refresh_token_expires_in": oauthToken.refreshTokenExpiresIn,
                "refresh_token_expired_at": oauthToken.refreshTokenExpiredAt,
                "id_token": oauthToken.idToken ?? "", // Can be nil
                "token_type": oauthToken.tokenType,
            ])
        }
    }

}
